Feature improve documentation (#329)
usnistgov/metaschema: website/content/specification/_index.md

For example, given field instances in the sequence `{ A, B, C, D }`, if `{ A, C }` are a collapse group, and `{ B, D }` are a different collapse group. Then `{ A, C }` would be ordered before `{ B, D }`, since `A` is ordered before `B`.

Note: Collapsing may affect the relative ordering of field instances. If two field instances are non-adjacent and their flags match, then the later field instance will be moved to be adjacent to the first. **Do not use the collapsible feature if maintaining field sequences is important to your use.**

An example a collapsible field definition might look like the following.

```xml
<define-assembly name="assembly">
  <root-name>assembly</root-name>
  <model>
    <define-field name="field" max-occurs="unbounded" collapsible="yes">
      <group-as name="fields" in-json="ARRAY"/>
      <define-flag name="flag-default-a" default="a"/>
      <define-flag name="flag-required" required="yes"/>
      <define-flag name="flag-optional"/>
    </define-field>
  </model>
</define-assembly>
```

> 
> An example set of content instances follow. XML is provided first to illustrate the use of uncollapsed fields, since this feature applies to JSON and YAML only. The JSON and YAML collapsed versions follow.
> 
> ```xml {linenos=table,hl_lines=[2,4]}
> <assembly>
>   <field flag-required="required 1">field-value-1</field>
>   <field flag-required="required 2">field-value-2</field>
>   <field flag-default-a="a" flag-required="required 1">field-value-3</field>
> </assembly>
> ```
> 
> The XML example above illustrates 2 fields, on lines 2 and 4, which will be collapsed together into the same *collapse group* in the resulting JSON and YAML conversion. This is because they have the same flag values: `a` for the flag named `flag-default-a`, `required 1` for the flag named `flag-required`, and no value for the flag named `flag-optional`. The field on line 3 would be in a second *collapse group*.
> 
> When converted to JSON or YAML, these fields will be collapsed based on their collapse grouping, resulting in the following content instances.
> 
> {{< tabs JSON YAML >}}
> {{% tab %}}
> ```json {linenos=table,hl_lines=["4-7"]}
> {
> "assembly": {
>   "fields": [
>     {
>       "flag-required": "required 1",
>       "STRVALUE": [ "field-value-1", "field-value-3"]
>     },
>     {
>       "flag-required": "required 2",
>       "STRVALUE": "field-value-2"
>     }
>   ]
> }
> ```
> {{% /tab %}}
> {{% tab %}}
> ```yaml {linenos=table,hl_lines=["4-7"]}
> ---
> assembly:
>   fields:
>   - flag-required: "required 1"
>     STRVALUE:
>     - "field-value-1"
>     - "field-value-3"
>   - flag-required: "required 2"
>     STRVALUE: "field-value-2"
> ```
> {{% /tab %}}
> {{% /tabs %}}
> 
> If the JSON or YAML instance is converted back to XML, the sequencing of the fields will change due to use of the collapsible feature as follows.
> 
> ```xml {linenos=table,hl_lines=["2-3"]}
> <assembly>
>   <field flag-required="required 1">field-value-1</field>
>   <field flag-required="required 1">field-value-3</field>
>   <field flag-required="required 2">field-value-2</field>
> </assembly>
> ```
> 
> In the example above, the fields from the first *collapse group* would appear before the fields from the second *collapse group*. This is because the resulting JSON and YAML instance has no way of indicating the original sequencing, so it must rely on the sequencing provided by the ordering of the *collapse groups* and the values within each group.
> 

Reviewer: Per discussion with others in the community call today, I would be amenable to collapsible being dropped from the specification and current version of Metaschema.

```suggestion
```

Author: Created issue #354 about this.

![Information model / Metaschema module relationship](im-module.svg)

An [*information model*](terminology/#information-model) is an abstract representation of information from a given information [*domain*](terminology/#domain). An *information model* is composed of a set of semantically well-defined information structures that each represent an [*information element*](terminology/#information-element) in the information domain.

The primary purpose of the Metaschema Framework is to support the structured expression of an information model, which is represented as a [*Metaschema module*](terminology/#metaschema-module). A Metaschema module is used to represent the whole or a part of a information model for a given information domain in an information-centric, format-neutral form.

A Metaschema module contains a collection of reusable [*definitions*](terminology/#definition) that each represent a given information element in an information model. Each definition contains documentation about the meaning (semantics), structure (syntax), and use of a given information element.

# Definitions, Instances, and Local Definitions

> Metaschema uses 3 types of definitions to represent information elements with different structural shapes. These different types of definitions are: [`define-flag`](#top-level-define-flag), [`define-field`](#top-level-define-field), or [`define-assembly`](#top-level-define-assembly). These definition types are used as building blocks of a Metaschema-based model.

Author: ```suggestion
Metaschema uses 3 types of definitions to represent information elements with different structural shapes: [`define-flag`](#top-level-define-flag), [`define-field`](#top-level-define-field), and [`define-assembly`](#top-level-define-assembly). These definition types are used as building blocks of a Metaschema-based model.
```

Flag and assembly definitions have child instances, which represent an edge between the containing definition and another definition. Thus, an *instance* makes use of another definition, typically by reference.

Both field and assembly definitions optionally allow the inclusion of one or more child *flag instances*.

> An assembly definition also has a model which contains a sequence of *model instances*. A model instance is an instance of a field or assembly.

Author: Let's go with the shorter version:
```suggestion
An assembly definition also has a model which contains a sequence of *model instances*, each model instance is an instance being either a field or assembly instance.
```

An optional sequence of [multiline markup](/specification/datatypes/#markup-multiline) used to provide additional supporting notes related to the Metaschema module.

A `<remarks>` element is typically used to including explanatory commentary of any kind.

For example:

```xml
<remarks>
  <p>This is an example model to describe the components of a computer.</p>
  <p>The "computer" element is the root element of this model.</p>
</remarks>
```

{{<callout>}}
As a general purpose element, the `remarks` element is also permitted to appear elsewhere in the Metaschema module model. Its scope of application is tied to the location of use in the document. Thus, the top-level remarks describe the entire metaschema, while remarks on an object definition describe the object definition.
{{</callout>}}

## XML `<namespace>`

The required `<namespace>` element is a [uniform resource identifier](https://www.rfc-editor.org/rfc/rfc3986) (URI) that identifies the [XML Namespace](https://www.w3.org/TR/xml-names/#sec-namespaces) to use for XML instances of the model.

All information objects defined in the Metaschema module will be assigned to this namespace when handling related XML data.

> {{<callout>}}

Author: #357 Done.

{{% tab %}}
```yaml
---
field:
  flag-b: "value"
```
{{% /tab %}}
{{% tab %}}
```xml
<field flag-b="value"/>
```
{{% /tab %}}
{{% /tabs %}}

### `<remarks>`

The optional `<remarks>` element provides a place to add notes related to the use of the definition. Remarks can be used to clarify the semantics of the definition in specific conditions, or to better describe how the definition may be more fully utilized within a model. 

The `<remarks>` element is optional and may occur multiple times.

### `<example>`

The optional `<example>` element is used to provide inline examples, which are intended to illustrate the use the information element implemented by the definition. Examples are provided in XML, which can then be automatically converted into other formats.

The `example` element is optional and may occur multiple times.

## top-level `<define-flag>`

A flag definition, represented by the `<define-flag>` element, is used to declare a reusable [flag](/specification/terminology/#flag) within a Metaschema module.

A flag definition provides the means to implement a simple, named [*information element*](/specification/terminology/#information-element) with a value.

{{<callout>}}

> Flag definitions are the primary leaf nodes in a Metaschema-based model. Flags are intended to represent granular particles of identifying and qualifying information.

Author: I find this language a bit cumbersome, but also I am not completely happy with what I wrote.

I'd rather not discuss fields or field values here if possible. I'd also like to avoid introducing parallel terminology such as leaf nodes / terminal nodes.

Perhaps we could collaborate on something?

> Flag definitions are the primary leaf nodes in a Metaschema-based model. Flags are intended to represent granular particles of identifying and qualifying information.

Reviewer 2: Point taken about not confusing with ideas better explained elsewhere.

I ended up going there because of the ambiguity raised by 'primary leaf node' - if primary, what's secondary and do we have tertiary as well? More precise might be to say that they typically represent the majority of leaf nodes? Part of the reason for 'primary' is that there is only one other kind -- but arguably of the two, a field value is 'primary' since it typically shows a data element of some kind.

We could start by striking 'primary' and then seeing how to communicate the idea better?

> Flag definitions are the primary leaf nodes in a Metaschema-based model. Flags are intended to represent granular particles of identifying and qualifying information.

Author: This content has been rewritten, so this thread is OBE.

### `@as-type`

The `@as-type` attribute declares the type of the flag's value. If not provided, the default value is `string`.

The `@as-type` attribute must have a value that corresponds to one of the [simple data types](#simple-data-types). As a result, markup in flag values is not permitted.

### `@default`

The `@default` attribute specifies the default value for the flag. When a flag is specified as an optional child of a `<define-field>` or `<define-assembly>`, this value should be considered set for a content instance if the flag is omitted in that instance.

### `<constraint>`

Constraints are [covered later](#define-flag-constraints) in this specification.

## top-level `<define-field>`

A field definition, represented by the `<define-field>` element, is used to declare a reusable [field](/specification/terminology/#field) within a metaschema module.

A field definition provides the means to implement a complex named [*information element*](terminology/#information-element) with a value and an optional set of [*flag*](#flag) instances.

{{<callout>}}

> A field is an edge node in a Metaschema-based model. Fields are typically used to provide supporting information for a containing [*assembly*](#assembly). The flag instances, typically characterize or identify the fields value. With optional use of flags, a field can be more or less complex, depending on the modeling need.

Author: I wanted specifically to state *identify*. *Quantify* is not the same concept. Can you explain the *quantify* change?

> A field is an edge node in a Metaschema-based model. Fields are typically used to provide supporting information for a containing [*assembly*](#assembly). The flag instances, typically characterize or identify the fields value. With optional use of flags, a field can be more or less complex, depending on the modeling need.

Reviewer 2: Many flags offer qualifications of the field or of its value - indeed arguably most. For example, prop/@name gives a name for a property (an assembly), and @ns qualifies (adds more contingent information) to the property or indeed in this case to the name of the property.

Another example would be `include-controls/@with-child-controls`, which qualifies the scope of the inclusion directive.

Maybe another term besides 'qualify' could get the same idea across better.

> A field is an edge node in a Metaschema-based model. Fields are typically used to provide supporting information for a containing [*assembly*](#assembly). The flag instances, typically characterize or identify the fields value. With optional use of flags, a field can be more or less complex, depending on the modeling need.

Author: This content has been rewritten, so this thread is OBE.

| [`@scope`](#scope) | `local` or `global` | optional | `global` |

Elements:

| Element | Data Type | Use      |
|:---       |:---       |:---      |
| [`<formal-name>`](#formal-name) | [`string`](/specification/datatypes/#string) | 0 or 1 |
| [`<description>`](#description) | [`markup-line`](/specification/datatypes/#markup-line) | 0 or 1 |
| [`<prop>`](#prop) | special | 0 to ∞ |
| [`<use-name>`](#naming-and-use-name) | [`token`](/specification/datatypes/#token) | 0 or 1 |
| [`json-key`](#) | special | 0 or 1 |
| [`json-value-key`](#json-value-key) or<br/>[`json-value-key-flag`](#json-value-key-flag) | special | 0 or 1 |
| [`flag`](#flag-instance-children) or<br/>[`define-flag`](#define-flag-inline-definition) | special | 0 or ∞ |
| [`<constraint>`](#define-flag-constraints) | special | 0 or 1 |
| [`<remarks>`](#remarks) | special | 0 or 1 |
| [`<example>`](#example) | special | 0 to ∞ |

The attributes and elements specific to the `<define-field>` are described in the following subsections. The elements and attributes common to all definitions are [defined earlier](#common-definition-metadata) in this specification.

### `@as-type`

> The optional `@as-type` attribute declares the type of the field's value. If not provided, the default value is `string`.

Author: This is related to flags. I don't want to confusing things by discussing flags.

How about just making this change?
```suggestion
The optional `@as-type` attribute declares the type of the field's value. If not provided, the default type is `string`.
```

      "STRVALUE": [ "field-value-1", "field-value-3"]
    },
    {
      "flag-required": "required 2",
      "STRVALUE": "field-value-2"
    }
  ]
}
```
{{% /tab %}}
{{% tab %}}
```yaml {linenos=table,hl_lines=["4-7"]}
---
assembly:
  fields:
  - flag-required: "required 1"
    STRVALUE:
    - "field-value-1"
    - "field-value-3"
  - flag-required: "required 2"
    STRVALUE: "field-value-2"
```
{{% /tab %}}
{{% /tabs %}}

If the JSON or YAML instance is converted back to XML, the sequencing of the fields will change due to use of the collapsible feature as follows.

```xml {linenos=table,hl_lines=["2-3"]}
<assembly>
  <field flag-required="required 1">field-value-1</field>
  <field flag-required="required 1">field-value-3</field>
  <field flag-required="required 2">field-value-2</field>
</assembly>
```

In the example above, the fields from the first *collapse group* would appear before the fields from the second *collapse group*. This is because the resulting JSON and YAML instance has no way of indicating the original sequencing, so it must rely on the sequencing provided by the ordering of the *collapse groups* and the values within each group.

### `@default`

> 

Reviewer 2: Did we discuss this feature? Where is it used?

Do we provide advice to implementers on how to process Metaschema-described data in the absence of schemas or schema bindings providing normative defaults? (I can imagine an XSLT that would provide such defaults but that's XSLT.)

(In OSCAL we have a nominal default for the `@ns` flag, but for no fields tmk?)

Author: @wendellpiez We did a while back. You had mixed feelings about this feature.

This allows for the declaration of the nominal default.

I'd like to keep schema bindings out of the discussion here. This would be discussed in specs around schema generation, which is out of scope here. Given this, what additional advice would you like to include here?

The `@default` attribute specifies the default value for the field. When a flag is specified as an optional child of a `<define-field>` or `<define-assembly>`, this value should be considered set for a content instance if the field is omitted in that instance.

A `@default`value MUST only be provided when the data type specified by the `@as-type` is a [simple data types](#simple-data-types). Specifying a `@default` value for a [markup data type](/specification/datatypes/#markup-data-types) MUST result in a Metaschema format error.

Implementations when writing content instances MAY omit writing default values in order to produce a more concise expression of the content.

### `<flag>` Instance Children

A field may have zero or more flag instance children.

See [flag instances](#flag-instances)

### `<define-flag>` Inline Definition

See [inline `<define-flag>`](#inline-define-flag).

### `<json-key>`

TODO: Specify this. Note assembly points to this section.

### JSON Value Keys

TODO: discuss use only when flags may be present.

> In XML, the value of a field appears as the text child of the field's element. In JSON and YAML, a property name is needed for the value. The `<json-value-key>` and `<json-value-key-flag>` elements provide a means to control the behavior of how this value is represented.

Reviewer 2: ```suggestion
In XML, the value of a field appears as the data content (text child or children) of the field's element. In JSON and YAML, a property name is needed for the value. The `<json-value-key>` and `<json-value-key-flag>` elements provide a means to control the behavior of how this value is represented.
```

(In XML, text *children* in cases where content is interrupted by comments or PIs.)

```xml
<title>Twenty thousand <!-- 20000 --> Leagues Under the Sea</title>
```

The value is `Twenty thousand  Leagues Under the Sea` (and note extra space).

Author: Indeed. Not sure how that is relevant here. Maybe comments should be discussed in the data types page? Want to suggest something there?

Reviewer 2: Yeah ... certainly not a relevant detail. Just trying to get past the imprecision of "text child" since it is not guaranteed to be so simple.

```suggestion
In XML, the value of a field appears as the textual data content of the field's element. In JSON and YAML, a property name is needed for the value. The `<json-value-key>` and `<json-value-key-flag>` elements provide a means to control the behavior of how this value is represented.
```

      <define-flag name="flag-default-a" default="a"/>
    </define-field>
  </model>
</define-assembly>
```

Would allow the following content in JSON and YAML.

{{< tabs JSON YAML >}}
{{% tab %}}
```json {linenos=table,hl_lines=[6]}
{
  "assembly": {
    "fields": [
      {
        "flag-default-a": "b",
        "value": "value1"
      }
    ]
  }
}
```
{{% /tab %}}
{{% tab %}}
```yaml {linenos=table,hl_lines=[5]}
---
assembly:
  fields:
  - flag-default-a: "b"
    value: "value1"
```
{{% /tab %}}
{{% /tabs %}}

#### `<json-value-key-flag>`

The `<json-value-key>` element can be declared to use the value of specific flag on the field as the property name for the field's value.

The `@flag-name` attribute MUST reference the [name](#name) of a flag on the field whose value is to be used as the property name.

This results in a property that is a combination of the referenced flag's value and the field's value.

For example:

TODO: complete this example.

## top-level `<define-assembly>`

An assembly definition, represented by the `<define-assembly>` element, is used to declare a reusable [assembly](/specification/terminology/#assembly) within a Metaschema module.

> An assembly definition provides the means to implement a complex, composite, named [*information element*](/specification/terminology/#information-element) that has no value. An assembly definition consists of an optional set of [*flags*](#flag) and an optional sequence of model instances, which are instances of assemblies and fields.

Reviewer 2: ```suggestion
An assembly definition provides the means to implement a complex, composite, named [*information element*](/specification/terminology/#information-element), with no value of its own, used to collect and organize other information elements. An assembly definition comprises a set of [*flags*](#flag), optional or required, and a sequence of model instances, i.e. instances of assemblies and fields, which may be optional, required or repeatable.
```

Author: Not a fan of *comprises* as it makes it read the a set of flags are made up of an assembly. Flags can't be both optional and required.

How about the following.

```suggestion
An assembly definition provides the means to implement a complex, composite, named [*information element*](/specification/terminology/#information-element) that collects and organizes other information elements, with no value of its own. 

An assembly definition consists of an optional set of [*flags*](#flag) and an optional sequence of [model instances](#model-instances).
```
The model instance section can cover the remaining details about model instances.

{{<callout>}}

> An assembly is a compositional node in a Metaschema-based model. Assemblies are typically used to represent complex data objects that combine multiple information elements together into a composite object representing a larger semantic concept. The flag instances, typically characterize or identify this composite object, while the model instances represent the information being composed.

Reviewer 2: ```suggestion
An assembly is a compositional node in a Metaschema-based model. Assemblies are typically used to represent complex data objects, combining multiple information elements together into a composite object representing a larger semantic concept. An assembly's flag instances will typically characterize or identify this composite object, while its model instances represent the information being composed.
```

1. A `<use-name>` is provided on the instance. The *effective name* is the value of the `<use-name>` element on the instance.
1. A `<use-name>` is provided on the definition. The *effective name* is the value of the `<use-name>` element on the definition.
1. No `<use-name>` is provided on the definition. The *effective name* is the value of the `@name` attribute on the definition.

For example:

```xml
<define-flag name="flag-a">
  <use-name>flag-b</use-name>
</define-flag>
<define-field name="field">
  <flag ref="flag-a">
    <use-name>flag-c</use-name>
  </flag>
</define-field>
```

In the example above, the *effective name* of the definition is `flag-c`. If the `<use-name>` was omitted on the instance, the *effective name* would be `flag-b`. If the `<use-name>` was also omitted on the definition, the *effective name* would be `flag-a`.

The following content is valid to the model above.

{{< tabs JSON YAML XML >}}
{{% tab %}}
```json
{
  "field": {
    "flag-c": "value"
  }
}
```
{{% /tab %}}
{{% tab %}}
```yaml
---
field:
  flag-c: "value"
```
{{% /tab %}}
{{% tab %}}
```xml
<field flag-c="value"/>
```
{{% /tab %}}
{{% /tabs %}}

> 

Reviewer 2: Do we want to say anything (or is this covered) about not permitting clashing effective names? I.e. siblings (defined together in a model) and their groups must all have distinct names.

- `yes` - Require the flag to be present in content. Content missing the flag and its value will be considered invalid.

## Model Instances

TODO: Continue from here

### Using cardinalities and `group-as`

The child `field` and `assembly` elements share the following common set of attributes:

- `@ref`(type: NCName, use: required): References the name of the corresponding `define-field` or `define-assembly`.
- `@min-occurs` (type: nonNegativeInteger, use: optional, default: 0): Indicates the minimum allowed occurance of the corresponding `define-field` or `define-assembly`.
- `@max-occurs` (type: positiveInteger or "unbounded", use: optional, default: 1): Indicates the maximum allowed occurance of the corresponding `define-field` or `define-assembly`. The value `unbounded` indicates the the referenced component can occur any number of times.

The `group-as` element is required if the `@max-occurs` attribute has a value greater than '1' or `unbounded` to provide additional information about how to handle the collection of data.

The `group-as` element has the following set of attributes:

> - `@name`(type: NCName, use: required): The grouping name to use in XML, JSON, and YAML. Use of this name is further clarified by the `@in-xml` attribute, when used.

Reviewer 2: ```suggestion
- `@name`(type: NCName, use: required): The grouping name to use in JSON, YAML and XML (when exposed). Use of this name is further clarified by the `@in-xml` attribute, when used.
```

("When exposed" b/c it won't show unless the 'in-xml='GROUPED'`)

- `@in-json` (type: special, use: optional, default: SINGLETON_OR_ARRAY): 

    In all cases, `@name` value is used as the property name.

    | Value | JSON Behavior |
    |:--- |:--- |
    | ARRAY | The child objects are to be represented as an array of objects. |
    | SINGLETON_OR_ARRAY | If a single object is provided, then the child will be an object, otherwise the child objects will be represented as an array of objects. |
    | BY_KEY | An itermediate object will be used as the child, with property names equal to the value of the referenced `define-field` or `define-assembly` component's flag as specified by the `@json-key` attribute on that component. See [using `@json-key`](#using-json-key). |

- `@in-xml` (type: special, use: optional, default: UNGROUPED): 

>     | Value | JSON Behavior |

Reviewer 2: ```suggestion
| Value | XML Behavior |
```

    |:--- |:--- |

>     | GROUPED | The child elements will be placed within a wrapper element with a localname equal to the value of the `@name` attribute. Each child element's localname will be the `@name` of the referenced component. |

Reviewer 2: ```suggestion
| GROUPED | The child elements will be placed within a wrapper element with a local name equal to the value of the `@name` attribute. Each child element's local name will be the `@name` of the referenced component. |
```

Also we're not saying 'effective name' here maybe we should?

>     | UNGROUPED | The `@name` attribute is ignored. Each child element's localname will be the `@name` of the referenced component. |

Reviewer 2: ```suggestion
| UNGROUPED | In XML, the components will appear without a grouping (wrapper) element with their own effective names; the `group-as/@name is ignored. |
```
